---
title: "Rcupcake: An R package for querying and analyzing biomedical data through the BD2K PIC-SURE RESTful API"
author:
- name: Alba Gutierrez-Sacristan
  affiliation: 
  - &dbmi Department of Biomedical Informatics, Harvard Medical School 
  email: alba_gutierrez@hms.harvard.edu
- name: Romain Guedj
  affiliation: *dbmi
- name: Gabor Korodi
  affiliation: *dbmi
- name: Jason Stedman
  affiliation: *dbmi
- name: Laura I. Furlong
  affiliation: 
    - &grib Research Programme on Biomedical Informatics (GRIB)
- name: Chirag J. Patel
  affiliation: *dbmi
- name: Isaac S. Kohane
  affiliation: *dbmi
- name: Paul Avillach
  affiliation: *dbmi
date: "`r doc_date()`"
package: "`r pkg_ver('Rcupcake')`"
bibliography: general_overview.bib
csl: biomed-central.csl
vignette: >
  %\VignetteIndexEntry{Rcupcake: An R package for querying and analyzing EHR data through the BD2K PIC-SURE RESTful API}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document2:
  toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
BD2K RESTful API is a programmatic interface that provides access to different data sources, making easier data accessibility, analysis reproducibility and scalability.The `Rcupcake` package contains functions to query different databases through the BD2K RESTful API in order to analyse demographic and phenotype variables and compute co-occurrence. according to patient age, gender and variation status if data available. A special focus is made on visualization of the results, providing a variety of representation formats such as networks, heatmaps and barplots (Table \@ref(tab:viz-opt)).


The `Rcupcake` package contains functions to query different databases through the BD2K RESTful API, a programmatic interface that provides access to different data sources, making easier data accessibility, analysis reproducibility and scalability. The `Rcupcake` package includes analysis functions to study demographic population, phenotype descriptive analysis, prevalence and co-occurrence, according to patient age, gender and variation status if data available. A special focus is made on visualization of the results, providing a variety of representation formats such as networks, heatmaps and barplots (Table \@ref(tab:viz-opt)).


## Background

In the era of big data and precision medicine, the number of databases containing biomedical information is increasing in an exponential way. Enabling experts to focus on their research questions rather than in the computation data merging, access and management is one of the biggest challenges nowadays. The `Rcupcake` package has been developed for this purpose.

The `Rcupcake` package allows to make queries to different databases through the PIC-SURE API, simplifying the data connection and analysis. Moreover, it allows its integration with other packages available in R to develop bioinformatics analysis workflows. 

The tasks that can be performed with `Rcupcake` package are the following:

  1. Query through the BD2K PIC-SURE API RESTful API in order to retrieve the data.
  2. Demographic descriptive analyses based on age and gender
  3. Phenotype descriptive analyses taking into account continuous and categorical variables. 
  4. Phenotype prevalence and co-occurrence analyses according to age, gender and any variation variable if available. 
  5. Extraction of patients with a defined pair of phenotypes. 

  
In the following sections the specific functions that can be used to address each of these tasks are presented.

## Installation 

The package \texttt{Rcupcake} is provided through GitHub. To install \texttt{Rcupcake} the user must type the following commands in an R session:

```{r bioC, message=FALSE, eval=FALSE}
library( devtools )
install_github( "hms-dbmi/Rcupcake" )
library( Rcupcake )
```

```{r knirt, message=FALSE, echo=FALSE}
library( knitr )
```

## S4 objects 

### `cupcakeData`

The `cupcakeData` object contains the clinical dataset retrieve by the initial query and shows a summary of the following characteristics:

  - The number of variations (N. Variations)
  - The number of phenotype variables (N. Phenotypes)
  - The number of patients (N. Patients)


```{r cupcakeDataObj1, echo=FALSE}
load(system.file("extdata", "RcupcakeExData.RData", package="Rcupcake"))
```

```{r cupcakeDataObj2}
RcupcakeExData
class( RcupcakeExData )
```

`cupcakeData` object is the output of `dataframe2cupcake` and `z.score` functions. It is used as input for `demographic.summary` and `phenotype.summary`  descriptive functions . This object comes with a series of functions to allow users to interact with the information retrieved. These functions are `n.phenotype`, `n.variation` and `extract`. 

The `n.phenotype` function returns the phenotype variables present in the dataset included in the object.
```{r nphenotype}
nphenotypeExample <- Rcupcake::n.phenotype( RcupcakeExData )
class( nphenotypeExample )
nphenotypeExample
```

The `n.variation` function returns the variation variables present in the dataset included in the object.
```{r nvariation}
nvariationExample <- Rcupcake::n.variation( RcupcakeExData )
class( nvariationExample )
nvariationExample
```


The `extract` function returns a formatted `data.frame` with the complete set of information contained in the object. 
```{r extract}
extractExample <- Rcupcake::extract( RcupcakeExData )
class( extractExample )
kable( head( extractExample ) )
```



### `cupcakeResults`

The `cupcakeResults` object contains the co-occurrence analysis results and a summary of the different measurements:

  - The minimum age of the patients in the dataset (Age Min)
  - The maximum age of the patients in the dataset (Age Max)
  - Patients gender (Gender)
  - variation, if it is selected (Variation)
  - Number of patients in the age and gender interval in the database
  - Number of patients in the age and gender interval with the variation select
  - The prevalence of the variation in the dataset
  - Odds ratio range 
  - Relative risk range
  - Phi value range
  - Number of resultant co-occurrences

```{r cupcakeResultsObj1, echo=FALSE}
load(system.file( "extdata", "RcupcakeExResult.RData", 
                  package="Rcupcake" ))
```

```{r cupcakeResultsObj2}
cupcakeResults
class( cupcakeResults )
```

`cupcakeResults` object is the output of the `co.occurrence` function and it is used as input for `pheno.prevalence`, `cooc.heatmap`and `cooc.network`  functions. 

The `cupcakeResults` object come with the `extract` function. The `extract` function allows the user to retrieve the data stored in the object. The `extract` function returns a formatted `data.frame` with the complete set of results obtained from the co-occurrence analysis.

# Retrieve data from the database through the BD2K RESTful API 
The first step in order to perform any analysis is extracting the data related to the patients of interest in the databse  stores in the BD2K RESTful API. The `Rcupcake` R package allows the user to retrieve the data from a database following 4 different steps.
   
  1. Start session
  2. Select the variables of interest. 
  3. Build the JSON query
  4. Run the query. 

## Starting session 
The first step consists on starting a session in the database of interest. To do it, the user can apply the `start.session` function.

The `start.session` function establishes a connection to the database of interest. It requires two arguments:

  - _url_: the url of the database of interest.
  - _apiKey_: the key to successfully access to the data.


The `start.session` function output is a message showing if the connection was or not successfully. If everything is correct the output message will be: _'Start Session: success'_, otherwise the next message will appear: _'Start Session: failed. Please revise your url and apiKey'_


```{r startSession, eval=FALSE}
sessionEx <- start.session( 
              url         = "https://nhanes.hms.harvard.edu/", 
              apiKey      = "YOURKEY"
              )
sessionEx
```


## Select the variables of interest
The second step will be retrie all the paths from the database where our variables of interest are located. This step will allow us to create the JSON query to retrieve the data. 

The `Rcupcake` R package allows the user to retrieve the paths from the database applying the `get.children` function.

The `get.children` function retrieves the paths of interest from the database using as input two different arguments:

  - _url_: the url of the database of interest.
  - _fieldname_: the parent path where the variables of interest are located.


The `get.children` function output is a vector of paths. This output will be used as input for the `my.query` function that generates the JSON query. 

```{r get.children, eval=FALSE}
nhanesPcbs <- get.children( 
              url         = "https://nhanes.hms.harvard.edu/", 
              fieldname   = "/nhanes/Demo/laboratory/laboratory/pcbs/"
              )
nhanesDemog <- get.children( 
              url         = "https://nhanes.hms.harvard.edu/", 
              fieldname   = "/nhanes/Demo/demographics/demographics/"
              )
```

## Build the JSON query
Once the paths are retrieved, the third step is to create the JSON query for the variables of interest. The `Rcupcake` R package allows the user to generate this query applying the `my.query` function.

The `my.query` function creates the JSON query using as input three different arguments:

  - _url_: the url of the database of interest.
  - _myfields_: the vector with the variables of interest. 
  - _myvector_:  the vector with the paths generated applying the `get.children` function


The `my.query` function output is the JSON query. This output will be used as input for the `my.query` function that generates the JSON query.

```{r my.query, eval=FALSE}
queryExample <- my.query( myfields = "AGE|PCB153", 
                          myvector = c(nhanesDemog, nhanesPcbs), 
                          url         = "https://nhanes.hms.harvard.edu/"
               )
```


## Run the query
The last step consists of retrieving the data. The `Rcupcake` R package allows the user to do it applying the `my.data` function. Given a url and a JSON query, it generates a `data.frame` object with the query output. 

The `my.data` function retrieves the data using as input three different arguments:

  - _url_: the url of the database of interest.
  - _query_: A JSON query, created applying the `my.query` function or contained in a text file.
  - _outputPath_:  the path and the file name where the output data will be saved. 


The `my.data` function output is the `data.frame` object that is retrieved for the JSON query of interest.

```{r my.data, eval=FALSE}
query <- my.data( 
              query  = system.file("extdata", "jsonQueryNhanes", package="Rcupcake"), 
              url    = "https://nhanes.hms.harvard.edu/"
              )
```   

# Transform the data in a cupcakeData object 
Once the data is in a data.frame format, it has to be transformed into a `cupcakeData` class object. Note that to perform a correct analysis, the data frame should contain:

  - 3 demographic variables: patient_id, Gender and Age. 
  - phenotype variables name
  - variation variables name
  
In order to transform the data.frame into a `cupcakeData` object, the `dataframe2cupcake` function has to be applied. The arguments needed as input are:

  - `input`: that determines the file with the complete path where the required input data.frame is located. 
  - `phenotypes`: vector that contains phenotype variables, separated by "|". 
  - `variants`: vector that contains the names of variations, separated by "|". 
  - `age`: vector that contains the age variable name. 
  - `gender`: vector that contains the gender variable name.
  
```{r convertToRcupcakeobject}
cupcakeExData <- dataframe2cupcake( input      = paste0(system.file("extdata", package="Rcupcake"), 
                                                "/queryOutput.txt"),
                                   age        = "Age",
                                   gender     = "Gender",
                                   phenotypes = "Diabetes|Arthritis|LiverCancer|AnyCancer",
                                   variants   = "PCB153",
                                   verbose    = TRUE)
```

As a result a `cupcakeData` object class is obtained. 
```{r convertToRcupcakeobject2}
class( cupcakeExData )
cupcakeExData 
head( Rcupcake::extract( cupcakeExData ) )
```

# Transform continuous in categorical variables: zscore
Phenotype variables can be continuous or categorical. Co-occurrence analysis will be only applied to those variables that are classified as categorical, like for example, `Diabetes`, which has `yes` or `no` values. If desired, the user can apply the `z.score` function for continuous variables. Given an object of class `cupcakeData`, the `z.score` function screens all the phenotype variables and transforms continuous into categorical variable applying Z-score. As a result a new `cupcakeData` object is generated. 

Note that if the number of individuals is lower than 5000 a Saphiro test [@shapiro1965analysis] will be done to test the normal distribution, otherwise a Kolmogorov-Smirnov test [@razali2011power] will be performed. The steps followed in this process are for each variable:

  1. Checking if the variable follows a normal distribution
  2. If the variable does not follow a normal distribution the Z-score will not be estimated.
  3. If the variable follows a normal distribution (p-value of shapiro or Kolmogorov-Smirnov test is lower than 0.05), then a correlation between the age and the variable y tested. When this correlation exists, a linear model is created. 
        
  4. Z-score will be estimated for this variable


As an input `z.score` function requires:

  * input: a `cupcakeData` object, obtained with the `dataframe2cupcake` function. 
  * zscoreCutOff: determines Z-score cut-off to categorize the continuous variable. By default it is set to -2 and 2. 
  * nfactor: determines the maximum number of values that a variable could have to be defined as categorical. By default it is 10. 


```{r zscore}
zscoreExample <- z.score( input        = cupcakeExData,
                          zscoreCutOff = c( -2, 2 ), 
                          nfactor      = 10, 
                          verbose      = TRUE )
```

As a result a new `cupcakeData` object is obtained. Note that the transformation cannot always be done. 

```{r zscoreExample}
head( Rcupcake::extract( cupcakeExData ) )
head( Rcupcake::extract( zscoreExample ) )
```

Note that the `z.score` function transforms the continuous phenotypes in categorical. In order to transform the variation, if necessary, the user should do it manually. For example, 'PCB153' variable, can be changed into categorical applying the code:

```{r variationTransform}
cupcakeExData@iresult$V.PCB153 <- ifelse( is.na(cupcakeExData@iresult$V.PCB153), "no", "yes" )
summary(as.factor(cupcakeExData@iresult$V.PCB153))
```

# Demographic analysis 
The `Rcupcake` R package allows the user to analyze and characterize the population under study. To have a general idea of the population main characteristics, the user can apply the `demographic.summary` function.

The function `demographic.summary` describes the demographic characteristics (sex, age) of the study population.
The `demographic.summary` function requires three arguments:

  - _input_: the `cupcakeData` object.
  - _maleCode_: the value which denotes males in the table (e.g., 0, M, Male...etc)
  - _femaleCode_: the value which denotes females in the table (e.g., 1, F, Female...etc)


The `demographic.summary` function output includes:
  
  - A barplot with the age distribution of the whole study population
  - A boxplot showing the age distribution by gender.
  - A pie chart representing the gender distribution.

```{r demographicAnalysis, fig.width=8, fig.height=8}
Rcupcake::demographic.summary ( input       = cupcakeExData, 
                               maleCode    = "male", 
                               femaleCode  ="female", 
                               verbose     = TRUE )
```


# Phenotype analysis
The `Rcupcake` R package includes the `phenotype.summary` function that describes the phenotypic characteristics for the whole study population by variation status.     

The `phenotype.summary` function requires as input: 

  - input: the `cupcakeData` object.
  - variation: the variation of interest (optional)
  - nfactor: determines the maximum number of values that a variable could have to be defined as categorical. By default it is 10. 
  - showTable: the table is displayed before the barplot. 
  - showFigures: results are visualized in barplot or boxplot, depending on the kind of variable. 
  - path: define the path where the output file generated will be saved. 

The `phenotype.summary` function output includes:   
  
  - A barplot for each categorical phenotype variable
  - A boxplot for each continuous phenotype variable
  - A table if the showTable argument is TRUE

```{r phenosumm, fig.width=8, fig.height=8, warning= FALSE}
phenotype.summary( input      = cupcakeExData,
                  showTable   = TRUE, 
                  showFigures = FALSE)
```


# Phenotype comparison 2 by 2
The user can compare each pair of variables in the input data by applying the `comparison2b2` function to the `cupcakeData` object previously generated with the `dataframe2cupcake` function. According to the type of variable - binary or continuous - different statistical tests.    


| Variable1 | Variable2 | Statistical test|
| ----------|-----------|-----------------|
| Binary    | Binary    | Fisher test     |
| Binary    | Continuous| T-test          |
| Continuous| Continuous| Pearson test    |
: (\#tab:statistics-opt) Statistical test performed according to the variables applying the `comparison2b2` function


The `comparison2b2` function requires as input: 

  - input: the `cupcakeData` object.
  - variable1: The name of the first variable that has to be the same that the colname of the variable in the input data.frame.
  - variable2: The name of the second variable that has to be the same that the colname of the variable in the input data.frame.
  - nfactor: determines the maximum number of values that a variable could have to be defined as categorical. By default it is 10. 

The `comparison2b2` function output shows the p-value of the statistical test performed and according to the variable type, some extra information is provided, - contingency table in the case of the fisher test and the mean and standard deviation in the case of the T-test. Three examples are provided. 

## Binary variables comparison

```{r comparison2b2Fisher, warning= FALSE}
comparison2b2( 
    input     = RcupcakeExData, 
    variable1 = "Gender",
    variable2 = "P.Diabetes"
)
```

## Continuous variables comparison

```{r comparison2b2Pearson, warning= FALSE}
comparison2b2( 
    input     = RcupcakeExData, 
    variable1 = "Age",
    variable2 = "V.PCB153"
)
```


## Binary vs continuous variables comparison
```{r comparison2b2Ttest, warning= FALSE}
comparison2b2( 
    input     = RcupcakeExData, 
    variable1 = "Gender",
    variable2 = "Age"
)
```

# Phenotypes co-occurrence 
The user can estimate the statistically significant co-occurrences by applying the `co.occurrence` function to the `cupcakeData` object previously generated with the `dataframe2cupcake` function. This function will return a new object (`cupcakeResults`) including co-occurrence measurements. Different functions can be applied to this object in order to show the phenotype prevalence, co-occurrence and phenotype plotting. 

## co-occurrence Measurements
The `co.occurrence` identifies all the possible phenotype co-occurrence and quantify them.  Five different quantification measures are estimated:   
  - Fisher test
  - co-occurrence score
  - Relative risk (RR)
  - Phi value (Pearson's correlation for binary variables)  
  - Odds ratio

### Fisher test
A Fisher exact test for each pair of diseases is performed to assess the null hypothesis of independence between two diseases. Four groups of patients are defined in order to perform the statistical testing: patients with phenotype A and phenotype B, patients with phenotype A but not phenotype B, patients with phenotype B but not phenotype A and patients without any of the two phenotypes. The Fisher exact test is then applied to estimated the p-value for each pair of diseases. The Benjamini-Hochberg false discovery rate method is applied on the ranked list to correct for multiple testing.
  
### co-occurrence score
This score is defined in Roque et al. [@roque2011using] as follows:$$comorbidity score =log_2 \left( \cfrac{observed + 1}{expected + 1} \right)$$
$$expected = \cfrac{n_{A} n_{B}}{N}$$

where _observed_ stands for the number of phenotype-phenotype associations (phenotype A and phenotype B), and _expected_ is estimated based on the occurrence of each phenotype (the number of patients diagnosed with phenotype A, _nA_, multiplied by the number of patients diagnosed with the co-occurrent phenotype B, _nB_ , and divided by the total number of patients, _N_). Since logarithm is applied, a co-occurrence score of 1.0 means that the observed co-occurrences are higher than two fold (approximately) than expected.
    
### Relative risk (RR)
The relative risk (RR) expresses the relationship between the rate of the prevalence of a phenotype among patients exposed and those patients that are not exposed to a certain risk factor. Here the risk factor is other phenotype. 
  * RR = 1: The risk is the same for patients exposed to the factor of risk than for those patients that are not exposed. 
  * RR > 1: The patients exposed to the factor of risk are more likely to suffer the disease. It is a risk factor.
  * RR < 1: The patients exposed to the factor of risk are less likely to suffer the disease. It is a protection factor.

The RR is estimated as the fraction between the number of patients diagnosed with both phenotypes and random expectation based on the phenotype prevalence [@hidalgo2009dynamic]. The RR of observing a pair of phenotypes A and B affecting the same patient is given by: $$RR_{AB} = \cfrac{C_{AB} N}{P_A P_B}$$

where _CAB_ is the number of patients affected by both phenotypes, _N_ is the total number of patients in the population and _PA_ and _PB_ are the prevalence of diseases A and B.

### Phi value (Pearsons correlation for binary variables) 
Phi value measures the robustness of the co-occurrence association. It can be expressed mathematically as:$$\phi_{AB} = \cfrac{C_{AB} N - P_{A} P_{B}}{\sqrt{P_{A} P_{B}(N - P_{A})(N - P_{A})}}$$

where _N_ is the total number of patients in the population, _PA_ i and _PB_ are the prevalence of phenotypes A and B respectively. _CAB_ is the number of patients that have been diagnosed with both phenotypes A and B, and _PA PB_ is the random expectation based on phenotype prevalence. The Pearson correlation coefficient, can take a range of values from +1 to -1. A value of 0 indicates that there is no correlation between two diseases; a value greater than 0 indicates a positive correlation between the two diseases and a value less than 0 indicates a negative correlation.

### Odds ratio
The odds ratio represents the increased chance that someone suffering phenotypes A will have the co-occurrent phenotypes B. It shows the extent to which suffering a phenotype increases the risk of developing another phenotype. The odds ratio is derived from a comparison of rates of the phenotypes among individuals who do and do not exhibit the factor of interest. A statistically significant odds ratio (significantly different from 1.00 at the .05 level) indicates an appreciable risk associated with a particular factor. For example, an odds ratio of 2.00 indicates a doubled risk of the appearance of the phenotype

** These measures allow the user to quantify the co-occurrence of phenotype pairs compared with the random expectation. The user can select the measure and the cut-off value in order to assess phenotype co-occurrence.**

The `co.occurrence` function requires 6 arguments:  

  - input: the `cupcakeData` object
  - pth: the path where the file with the phenotype values generated previously is located.
  - aggregate: if all the possible phenotype values want to be considered individually, aggregate must be FALSE. otherwise, the phenotype values should be manually completed by the user (column yes/no) and aggregate must be TRUE. 
  - ageRange: determines what is the age range of interest for performing the analysis.
  - gender: determine what is the gender of interest to perform the co-occurrence analysis. 
  - variation: determine the variation of interest and the value of this variation.


```{r phenoCooccurrence}
cupcakeResults <- co.occurrence( 
              input         = cupcakeExData,
              pth           = system.file("extdata", package="Rcupcake"), 
              aggregate    = FALSE, 
              ageRange      = c(0,85),
              gender        = "ALL"
              )
```

As a result, a `cupcakeResults` object is obtained. This object contains a summary of the phenotype co-occurrences that have been found.

```{r phenoCooccurrence2}
class( cupcakeResults )
cupcakeResults
```


## Plotting the co-occurrence results

`Rcupcake` offers several options to visualize the results from the phenotype co-occurrence analysis, in networks and heatmaps.

### Network
The `cooc.network` function returns a network showing phenotype co-occurrences. By default `Rcupcake` package shows a network with the `layout.fruchterman.reingold` layout, where the phenotypes are joined according to the number of patients that have both phenotypes ( selectValue = patientsPhenoAB). The user can change this value to the different co-occurrence measures - fisher, oddsRatio, relativeRisk, phi, expect, score, fdr, PercentagePhenoAB- that are estimated for each pair of phenotypes. 

The `cooc.network` function requires 3 arguments as input:  

   - input: the `cupcakeResults` object
   - representedVariable: the co-occurrence measurement 
   - variableCutOff: the numeral value of the cutOff

The `cooc.network` function output is a network. Node size is proportional to the phenotype prevalence. Edge numbers represent the selected value. 


```{r phenoNetwork, fig.width=10, fig.height=10}
cooc.network ( input               = cupcakeResults,
               representedVariable = "patientsPhenoAB", 
               variableCutOff      = 100)
```


### Heatmap
The `cooc.heatmap` function returns a heatmap. This function is homologous to `coo.network` one.  The heatmap represents the selected value that by default will be `patientsPhenoAB`. The user can change this value to the different co-occurrence measures - fisher, oddsRatio, relativeRisk, phi, expect, score, fdr, PercentagePhenoAB- that are estimated for each pair of phenotypes. 

The `cooc.heatmap` function requires 3 arguments:  

  - input: the `cupcakeResults` object
  - representedVariable: the co-occurrence measurement 
  - variableCutOff: the numeral value of the cutOff

The `cooc.heatmap` function output is a heatmap. Blue color represents the lower values,  the yellow color represents the upper values. The co-occurrence measurement value is showed in each cell. 

```{r phenoHeatmap, fig.width=10, fig.height=10 }
cooc.heatmap ( input               = cupcakeResults,
               representedVariable = "patientsPhenoAB", 
               variableCutOff      = 100)
```

## Plotting the prevalence results

```{r phenoPrevalenceNetwork, fig.width=10, fig.height=10 }
prevalence.plot( input   = cupcakeResults,
                 variation = "PCB153",
                   verbose = TRUE
              )
```


If no variation is indicated, the output of the `prevalence.plot` will be a barplot showing the general prevalence of each phenotype in the data set. 

```{r phenoPrevalenceBarplot, fig.width=10, fig.height=10 }
prevalence.plot( input   = cupcakeResults,
                   verbose = TRUE
              )
```


# Summary of `Rcupcake` functions available

| Input Object    | `Rcupcake` function | Output Generated                                  |
| ----------------|-----------------------|---------------------------------------------------|
| -               | `my.data`             | Given an url, a key and a JSON object, it generates a data.frame object with the output of the query|
|-                | `dataframe2cupcake`           | Given a tabulated file, checks if it contains the data in the correct format and generates a `cupcakeData` objec|
|
| `cupcakeData`     | `demographic.summary`  | 3 plots: age and gender distribution, relation between age and gender distribution|
|                 | `phenotype.summary`     | A data.frame with the prevalence of each phenotype according to the value it takes, is retrieved. A figure showing the results can be also obtained. |
|                 | `extract`             | It generates a data.frame containing the raw data from a query |
|                 | `z.score`     | It transforms continuous into categorical variable applying Z-score |
|                 | `comparison2b2`   | It performs a statistical analysis test to check the relation between two given variables present in the cupcakeData object |
|                 | `co.occurrence`| It generates a cupcakeResults object              |
|                 | `patient.selection`| It retrieves a list of patients suffering two phenotypes of interest |
|                 | `n.variation`           | It retrieves the variations of your data |
|                 | `n.phenotype`           | It retrieves the phenotypes of your data |
| cupcakeResults  | `cooc.heatmap`    | It generates a heatmap showing the co-occurrence analysis results |
|                 | `cooc.network`    | It generates a network showing the co-occurrence analysis results |
|                 | `extract`             | It generates a data.frame containing the co-occurrence results |
|                 | `pheno.prevalence` | It generates a data.frame containing the prevalence of each phenotype |
|                 | `prevalence.plot`      | It generates a barplot or a network showing the results of the  phenotype's prevalence |

: (\#tab:viz-opt) Functions in `Rcupcake` R package



# Bibliography
